File: src/runtime/cpp/std/pathlib.cpp
#include "pathlib.h"

#include <stdexcept>

#include "glob.h"
#include "std/os.h"
#include "std/os_path.h"

Path::Path(const str& value) : _value(value) {}

Path::Path(const char* value) : _value(value == nullptr ? "" : value) {}

Path::Path(const object& value) : _value(str(py_to_string(value))) {}

str Path::__str__() const {
    return _value;
}

str Path::__repr__() const {
    return str("Path(") + _value + str(")");
}

str Path::__fspath__() const {
    return _value;
}

Path Path::__truediv__(const object& rhs) const {
    return Path(join(_value, str(py_to_string(rhs))));
}

Path Path::parent() const {
    str parent_txt = dirname(_value);
    if (parent_txt == "") {
        parent_txt = str(".");
    }
    return Path(parent_txt);
}

list<Path> Path::parents() const {
    list<Path> out{};
    str current = dirname(_value);
    while (true) {
        if (current == "") {
            current = str(".");
        }
        out.append(Path(current));
        str next_current = dirname(current);
        if (next_current == "") {
            next_current = str(".");
        }
        if (next_current == current) {
            break;
        }
        current = next_current;
    }
    return out;
}

str Path::name() const {
    return basename(_value);
}

str Path::suffix() const {
    return ::std::get<1>(splitext(basename(_value)));
}

str Path::stem() const {
    return ::std::get<0>(splitext(basename(_value)));
}

Path Path::with_suffix(const str& suffix_value) const {
    return Path(::std::get<0>(splitext(_value)) + suffix_value);
}

Path Path::relative_to(const object& other) const {
    str base = str(py_to_string(other));
    str self_abs = abspath(_value);
    str base_abs = abspath(base);
    if (!base_abs.empty() && base_abs.std().back() != '/') {
        base_abs += str("/");
    }
    if (self_abs == base_abs || (!base_abs.empty() && self_abs == base_abs.substr(0, base_abs.size() - 1))) {
        return Path(str("."));
    }
    const ::std::string self_std = self_abs.std();
    const ::std::string base_std = base_abs.std();
    if (self_std.rfind(base_std, 0) == 0) {
        return Path(str(self_std.substr(base_std.size())));
    }
    throw ::std::runtime_error(self_abs.std() + " is not relative to " + base.std());
}

Path Path::resolve() const {
    return Path(abspath(_value));
}

bool Path::exists() const {
    return ::exists(_value);
}

void Path::mkdir(bool parents_flag, bool exist_ok) const {
    if (parents_flag) {
        makedirs(_value, exist_ok);
        return;
    }
    if (exist_ok && ::exists(_value)) {
        return;
    }
    ::mkdir(_value, exist_ok);
}

str Path::read_text(const str& encoding) const {
    (void)encoding;
    auto file = open(_value, str("r"));
    return file.read();
}

int64 Path::write_text(const str& text, const str& encoding) const {
    (void)encoding;
    auto file = open(_value, str("w"));
    return file.write(text.std());
}

Path Path::joinpath(const str& part) const {
    return Path(join(_value, part));
}

Path Path::joinpath(const Path& part) const {
    return Path(join(_value, part._value));
}

Path Path::joinpath(const str& part0, const str& part1) const {
    return Path(join(join(_value, part0), part1));
}

list<Path> Path::glob(const str& pattern) const {
    list<Path> out{};
    for (const auto& item : ::glob(join(_value, pattern))) {
        out.append(Path(item));
    }
    return out;
}

Path Path::cwd() {
    return Path(getcwd());
}

File: src/runtime/julia/std/pathlib.jl
include(joinpath(@__DIR__, "os_native.jl"))
include(joinpath(@__DIR__, "os_path_native.jl"))

mutable struct Path
    _value
    Path() = new("")
end

function Path(value)
    p = Path()
    p._value = string(value)
    return p
end

function cwd()
    return Path(__OsNative.getcwd())
end

Base.show(io::IO, p::Path) = print(io, p._value)
Base.showerror(io::IO, p::Path) = print(io, p._value)

function Base.getproperty(p::Path, name::Symbol)
    if name === :name
        return __OsPathNative.basename(getfield(p, :_value))
    end
    if name === :stem
        root, _ext = __OsPathNative.splitext(__OsPathNative.basename(getfield(p, :_value)))
        return root
    end
    if name === :parent
        return Path(__OsPathNative.dirname(getfield(p, :_value)))
    end
    return getfield(p, name)
end

function Base.:/(lhs::Path, rhs)
    return Path(__OsPathNative.join(getfield(lhs, :_value), string(rhs)))
end

function Base.joinpath(lhs::Path, rhs...)
    parts = Any[getfield(lhs, :_value)]
    for part in rhs
        push!(parts, string(part))
    end
    return Path(__OsPathNative.join(parts...))
end

function exists(self::Path)
    return __OsPathNative.exists(self._value)
end

function mkdir(self::Path, parents=false, exist_ok=false)
    if parents
        return __OsNative.makedirs(self._value, exist_ok)
    end
    return __OsNative.mkdir(self._value, exist_ok)
end

function mkdir(self::Path; parents=false, exist_ok=false)
    return mkdir(self, parents, exist_ok)
end

function write_text(self::Path, text)
    open(self._value, "w") do io
        write(io, string(text))
    end
    return nothing
end

function write_text(self::Path, text; encoding="utf-8")
    open(self._value, "w") do io
        write(io, string(text))
    end
    return nothing
end

function read_text(self::Path)
    return read(self._value, String)
end

function read_text(self::Path; encoding="utf-8")
    return read(self._value, String)
end
